### PR TITLE
[DL-518] flow io payments captures

### PR DESCRIPTION
**Pull request by sebastiandl**
### What problem is the code solving?
When an order is completed through the Flow's checkout, it is redirected back to Spree, where Spree perform a series of actions like pulling data from Flow's api, creating records based on that information and finally redirecting the user to the Thank you page. At this point, the order's payment is left as **pending** until Spree receives the **capture_upserted_v2** webhook. Unfortunately, there are cases where the webhook is triggered almost instantly after the user has completed the checkout, and the order is incomplete.

### How does this change address the problem?
- Updating the **capture_upserted_v2** webhook to validate if order has payments present. If it does it will process the capture and If not it will schedule a job to do it. because of this I moved some code into a new method that would be call from within the job as well.
- Creating a new job **update_payment_capture_worker.rb** that would process the capture. This worker is configured to try processing the capture up to three times. If there is a problem while running it will notify us via Slack.
- Moving shared code between workers inside a new file **flow_io_worker.rb** so it can be shared between Flow workers.

### Why is this the best solution?
Unfortunately, we don't have too many alternatives. We could set some kind of cron or different job to pull this data from Flow's api, but Flow should be the one telling us when the payment is finally capture, so we kind of have to rely on the webhook.

One improvement to be done in the near futre is moving the code related to the upserting the captures outside the Webhook method and into probably the OrderUpdater class. I tried doing it in this ticket but I realized it was going to take me lots of time. I will create a ticket to improve this and also include a little code to pull the captures data from the Flow's api if needed.

Summing up, this was the simplest and more direct option I could think ok.

### Share the knowledge
When testing this, we need to have at least two webservers running (or start a webserver with multiple workers probably). In order to do that we could either update the scalability configuration of an environment to have two instances or we could update the nginx configuration to load multiple workers.

Once we have that, we simply need to start creating orders in Flow paying with Paypal. These orders should have their payment updated as 'completed' and they should have the capture event created as well.

### CODE REVIEW

File: app/workers/flowcommerce_spree/flow_io_worker.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    sidekiq_retries_exhausted do |message, exception|
+      Rails.logger.warn("[!] #{self.class} max attempts reached: #{message} - #{exception}")
+      notification_setting = FlowcommerceSpree::Config.notification_setting
+      return unless notification_setting[:slack].present?
+
+      slack_message = "[#{Rails.env}] #{message}"
+      Slack_client.chat_postMessage(channel: notification_setting[:slack][:channel], text: slack_message)
+    end
+  end
+end
```
Comment (earthtodani) [inline app/workers/flowcommerce_spree/flow_io_worker.rb:16]: ❤️

**Review by earthtodani — APPROVED** [review]

**Comment by earthtodani** [comment]
> In order to be able to test if this works, we need an environment with at least two webservers running at the same time.

Could you explain a bit more of the testing strategy?

**Comment by sebastiandl** [comment]
> > In order to be able to test if this works, we need an environment with at least two webservers running at the same time.
> 
> @sebastiandl Could you explain a bit more of the testing strategy?

Yes! The problem we are having is that Flow redirects the user to the Thank you page, and immediately after triggers the payment capture event for some payment methods. In environments where we only have one webserver running, these calls do not run at the same time. But in environments like production or staging where we have multiple instances running there is a huge change that these run at the same time in different instances.

In order to replicate the behavior, we have to at least have two webservers running (or one webserver with multiple workers). In order to do that we could either update the scalability configuration of an environment to have two instances or we could update the nginx configuration to load multiple workers probably.

I will update the description with this.

**Review by MarsiliFrancoDavid — APPROVED** [review]
🙇

File: app/services/flowcommerce_spree/webhooks/capture_upserted_v2.rb [hunk 2]
```diff
@@ -22,16 +22,25 @@ def process
...
 
         if (order = Spree::Order.find_by(number: order_number))
-          upsert_order_captures(order, capture)
-          payments = order.flow_io_payments
-          map_payment_captures_to_spree(order, payments) if payments.present?
-          order
+          if order.payments.any?
+            store_payment_capture(order, capture)
+          else
+            FlowcommerceSpree::UpdatePaymentCaptureWorker.perform_in(1.minute, order.number, capture)
```
Comment (maurosettier) [inline app/services/flowcommerce_spree/webhooks/capture_upserted_v2.rb:28]: this minute is the ultimate time? this time will always be enough?

Comment (sebastiandl) [reply app/services/flowcommerce_spree/webhooks/capture_upserted_v2.rb:28]: Sorry @maurosettier for the delay!! Yes, it should be enough! This method generally does not take more than seconds. If there is a problem though (timeout, bad gateway. etc) We would end up getting informed in the Slack channel `#flow-io-alerts` thanks to how we set up the DelayedJob!

**Review by alejandc — APPROVED** [review]
